File: UserPortal/electron/kioskServer.js
const express = require('express');
const http = require('http');
const https = require('https');
const ws = require('ws');
const path = require('path');

// flutterWebDir is passed from main.js and resolves to the correct location
// in both dev (source tree) and packaged builds (resources/web/).

const SERVER_HOST = 'localhost';
const SERVER_PORT = 3080;
const KIOSK_SECRET = process.env.KIOSK_SECRET ?? 'dev-kiosk-secret';

/**
 * Forwards a request to the main HTTPS server at /api/kiosk/* and pipes the
 * response back. Injects the X-Kiosk-Secret header automatically.
 */
function proxyToServer(req, res) {
  const body = JSON.stringify(req.body ?? {});
  const targetPath = '/api/kiosk' + (req.url || req.path);

  const options = {
    hostname: SERVER_HOST,
    port: SERVER_PORT,
    path: targetPath,
    method: req.method,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(body),
      'X-Kiosk-Secret': KIOSK_SECRET,
    },
    rejectUnauthorized: false, // server uses self-signed cert
  };

  const proxyReq = https.request(options, (proxyRes) => {
    res.status(proxyRes.statusCode);
    proxyRes.pipe(res, { end: true });
  });

  proxyReq.on('error', (err) => {
    console.error('[KioskServer] Proxy error:', err.message);
    res.status(502).json({ error: 'Main server unavailable' });
  });

  proxyReq.write(body);
  proxyReq.end();
}

/**
 * Broadcasts a card event to all connected WebSocket clients.
 */
function makeBroadcaster(wsServer) {
  return function broadcast(type, cardId) {
    const message = JSON.stringify({ type, cardId });
    wsServer.clients.forEach((client) => {
      if (client.readyState === ws.WebSocket.OPEN) {
        client.send(message);
      }
    });
  };
}

/**
 * Starts the local Express + WebSocket server that:
 *   - Serves the Flutter web build as a static SPA
 *   - Proxies /kiosk-api/* to the main HTTPS server
 *   - Broadcasts card scan events over WebSocket to Flutter
 *
 * @param {{ doorCardReader, toolCardReader, port: number }} options
 * @returns {http.Server}
 */
function startKioskServer({ doorCardReader, toolCardReader, port, flutterWebDir }) {
  const app = express();
  app.use(express.json());

  // Dev-only: simulate card scans without hardware
  if (process.env.NODE_ENV !== 'production') {
    app.post('/kiosk-api/dev/simulate-door-scan', (req, res) => {
      const cardId = req.body?.cardId ?? 'AABBCCDD';
      doorCardReader.simulateScan(cardId);
      res.json({ ok: true, cardId });
    });

    app.post('/kiosk-api/dev/simulate-tool-scan', (req, res) => {
      const cardId = req.body?.cardId ?? '11223344';
      toolCardReader.simulateScan(cardId);
      res.json({ ok: true, cardId });
    });
  }

  // Proxy all other /kiosk-api/* requests to the main server
  app.use('/kiosk-api', proxyToServer);

  // Serve Flutter web build
  app.use(express.static(flutterWebDir));

  // SPA fallback: Flutter handles client-side routing
  app.get('*', (_req, res) => {
    res.sendFile(path.join(flutterWebDir, 'index.html'));
  });

  const server = http.createServer(app);

  // WebSocket server for card scan events
  const wsServer = new ws.WebSocketServer({ server, path: '/card-events' });
  const broadcast = makeBroadcaster(wsServer);

  wsServer.on('connection', () => {
    console.log('[KioskServer] Flutter connected via WebSocket');
  });

  // Forward hardware events to all connected Flutter clients
  doorCardReader.on('card', ({ cardId }) => {
    console.log('[KioskServer] Door card scanned:', cardId);
    broadcast('door-card', cardId);
  });

  toolCardReader.on('card', ({ cardId }) => {
    console.log('[KioskServer] Tool card scanned:', cardId);
    broadcast('tool-card', cardId);
  });

  return new Promise((resolve) => {
    server.listen(port, () => {
      console.log(`[KioskServer] Listening on http://localhost:${port}`);
      resolve(server);
    });
  });
}

module.exports = { startKioskServer };

File: UserPortal/electron/kiosk_server.py
"""
Local kiosk server — Python/FastAPI equivalent of kioskServer.js.

Provides:
  - HTTP proxy: /kiosk-api/* → main HTTPS server at TOOLACCESS_SERVER:3080
  - WebSocket:  /card-events  broadcasts card scan events to Flutter
  - Static:     serves Flutter web build when flutter_web_dir is supplied
  - Dev:        /kiosk-api/dev/simulate-{door,tool}-scan (non-production only)
"""

import asyncio
import json
import os
from pathlib import Path
from typing import Optional, Set

import httpx
from fastapi import FastAPI, Request, WebSocket, WebSocketDisconnect
from fastapi.responses import FileResponse, JSONResponse, Response

SERVER_HOST = os.environ.get('TOOLACCESS_SERVER', 'localhost')
SERVER_PORT = 3080
KIOSK_SECRET = os.environ.get('KIOSK_SECRET', 'dev-kiosk-secret')
IS_DEV = os.environ.get('KIOSK_ENV', 'development') != 'production'

# Shared state
_ws_clients: Set[WebSocket] = set()
_http_client: Optional[httpx.AsyncClient] = None
_loop: Optional[asyncio.AbstractEventLoop] = None


async def _broadcast(event_type: str, card_id: str) -> None:
    """Send a card event JSON message to every connected WebSocket client."""
    message = json.dumps({'type': event_type, 'cardId': card_id})
    disconnected: Set[WebSocket] = set()
    for ws in _ws_clients:
        try:
            await ws.send_text(message)
        except Exception:
            disconnected.add(ws)
    _ws_clients.difference_update(disconnected)


def create_app(door_card_reader, tool_card_reader, flutter_web_dir: Optional[str] = None) -> FastAPI:
    app = FastAPI()

    # ------------------------------------------------------------------ lifecycle

    @app.on_event('startup')
    async def on_startup() -> None:
        global _http_client, _loop
        _loop = asyncio.get_running_loop()
        # verify=False mirrors rejectUnauthorized: false — main server uses self-signed cert
        _http_client = httpx.AsyncClient(verify=False)

    @app.on_event('shutdown')
    async def on_shutdown() -> None:
        if _http_client:
            await _http_client.aclose()

    # ------------------------------------------------------------------ dev endpoints
    # Must be registered before the /kiosk-api/{path:path} catch-all.

    if IS_DEV:
        @app.post('/kiosk-api/dev/simulate-door-scan')
        async def simulate_door_scan(request: Request) -> JSONResponse:
            body: dict = {}
            try:
                body = await request.json()
            except Exception:
                pass
            card_id: str = body.get('cardId', 'AABBCCDD')
            door_card_reader.simulate_scan(card_id)
            return JSONResponse({'ok': True, 'cardId': card_id})

        @app.post('/kiosk-api/dev/simulate-tool-scan')
        async def simulate_tool_scan(request: Request) -> JSONResponse:
            body: dict = {}
            try:
                body = await request.json()
            except Exception:
                pass
            card_id: str = body.get('cardId', '11223344')
            tool_card_reader.simulate_scan(card_id)
            return JSONResponse({'ok': True, 'cardId': card_id})

    # ------------------------------------------------------------------ proxy

    @app.api_route('/kiosk-api/{path:path}', methods=['GET', 'POST', 'PUT', 'DELETE', 'PATCH'])
    async def proxy(path: str, request: Request) -> Response:
        """Forward requests to the main HTTPS server, injecting X-Kiosk-Secret."""
        body = await request.body() or b'{}'
        target_url = f'https://{SERVER_HOST}:{SERVER_PORT}/api/kiosk/{path}'
        try:
            resp = await _http_client.request(
                method=request.method,
                url=target_url,
                content=body,
                headers={
                    'Content-Type': 'application/json',
                    'X-Kiosk-Secret': KIOSK_SECRET,
                },
            )
            return Response(
                content=resp.content,
                status_code=resp.status_code,
                media_type='application/json',
            )
        except Exception as exc:
            print(f'[KioskServer] Proxy error: {exc}')
            return JSONResponse({'error': 'Main server unavailable'}, status_code=502)

    # ------------------------------------------------------------------ websocket

    @app.websocket('/card-events')
    async def card_events(websocket: WebSocket) -> None:
        await websocket.accept()
        _ws_clients.add(websocket)
        print('[KioskServer] Flutter connected via WebSocket')
        try:
            while True:
                await websocket.receive_text()  # keep connection alive
        except WebSocketDisconnect:
            pass
        finally:
            _ws_clients.discard(websocket)

    # ------------------------------------------------------------------ static files (Flutter web)

    if flutter_web_dir:
        flutter_path = Path(flutter_web_dir)

        @app.get('/{full_path:path}')
        async def serve_flutter(full_path: str) -> FileResponse:
            candidate = flutter_path / full_path
            if candidate.is_file():
                return FileResponse(str(candidate))
            return FileResponse(str(flutter_path / 'index.html'))

    # ------------------------------------------------------------------ hardware events → websocket

    def _make_card_handler(event_type: str):
        def handler(event: dict) -> None:
            card_id: str = event['cardId']
            print(f'[KioskServer] {event_type} scanned: {card_id}')
            if _loop and _loop.is_running():
                # Safe to call from both the event-loop thread and background hardware threads.
                _loop.call_soon_threadsafe(
                    asyncio.ensure_future,
                    _broadcast(event_type, card_id),
                )
        return handler

    door_card_reader.on('card', _make_card_handler('door-card'))
    tool_card_reader.on('card', _make_card_handler('tool-card'))

    return app
